3.0/vector_tile.proto

```diff
@@ -0,0 +1,123 @@
+                optional int64 int_value = 4;
+                optional uint64 uint_value = 5;
+                optional sint64 sint_value = 6;
+                optional bool bool_value = 7;
+
+                // Even numbered elements refer to the nth key in the layer's keys list
+                // Odd numbered elements refer to the nth value in the layer's values list
+                repeated uint32 hash_value = 8 [ packed = true ];
+
+                // Each element refers to the nth value in the layer's values list
+                repeated uint32 list_value = 9 [ packed = true ];
+
+                extensions 10 to max;
+        }
+
+        // Features are described in section 4.2 of the specification
+        message Feature {
+                // Use object_id if the id is not an integer
+                optional uint64 id = 1;
+
+                // Tags of this feature are encoded as repeated pairs of
+                // integers.
+                // A detailed description of tags is located in sections
+                // 4.2 and 4.4 of the specification
+                repeated uint32 tags = 2 [ packed = true ];
+
+                // The type of geometry stored in this feature.
+                optional GeomType type = 3 [ default = UNKNOWN ];
+
+                // Contains a stream of commands and parameters (vertices).
+                // A detailed description of geometry encoding is located in
+                // section 4.3 of the specification.
+                repeated uint32 geometry = 4 [ packed = true ];
+
+                // Representation of spline knots to be clarified:
+                // https://github.com/mapbox/vector-tile-spec/issues/114
+                repeated double knots = 5 [ packed = true ];
+
+                // If present, marks the geometry as explicitly MultiPoint,
+                // MultiLineString, or MultiPolygon
+                optional bool multigeometry = 6;
+
+                // If present, these are references into the layer's values,
+                // one per node (excluding ClosePath) of the geometry,
+                // either to a hash_value giving the keys and values of
+                // the node's attributes, or to a null value, indicating
+                // that the node has no attributes.
+                repeated uint32 node_attributes = 7 [ packed = true ];
+
+                // If present, these are delta-encoded elevations for each
+                // node (excluding ClosePath) in the geometry. If any nodes have only
+                // two dimensions specified, the missing elevations must
+                // be encoded here as NaN. Only finite elevations participate
+                // in delta encoding: the delta of the elevation after
+                // a NaN is relative to the last finite elevation.
+                // Elevations are in meters above or below the WGS84 ellipsoid.
+                repeated double node_elevations = 8 [ packed = true ];
```

Comment:
A double here would take up quite a bit of space as it could not be delta encoded and would not use integers.

Reply:
We could do a custom byte stream that has separate type markers for integer delta, floating deltas, and resets to absolute values periodically to address drift, but that is getting very complicated unless there is an existing spec for this kind of thing that we could point to.

Reply:
### What unspecified elevation means

I guess my feeling was that 2D and 3D features could coexist, but that 2D features all implicitly had an elevation of 0. But that contracts GeoJSON, which says that coordinates are WGS84 but also that "In the absence of elevation values, applications sensitive to height or depth SHOULD interpret positions as being at local ground or sea level."

Even if 2D and 3D can't be in the same layer, it has to be possible to render them together, so they have to be reconciled somehow. A tile-level declaration of local ground level relative to the ellipsoid could work, but just moves the responsibility of knowing the local ground level everywhere in the world to the tile creator.

Maybe we could hedge and say that null/unspecified elevations are extruded to the highest intersecting point in any other features they are being rendered together with, so if there is a terrain layer, they inherit from it, but if there isn't, they are on the ellipsoid?

### Data type

The precision scale on integer elevations sounds like it would work (aside from worries above about intermixing known and unknown elevations) and would keep the math exact.

I do think it should be possible to have both specified and unspecified elevations within the same feature, because if you don't, you end up with data sets assigning magic values like "-999" to unknown elevations. So I think we should provide our own official magic value for explicit nulls within otherwise-3D features.

Reply:
I think it would unreasonable to expect that encoders and decoders know the actual elevation of the ground they are currently on in order to calculate the position of a point in 3d space. This would violate the concept of VTs being a single source of data if they had to reference another set of data to function. 

Additionally, I am still not completely convinced that elevation data should be different from the 2D data and could be all encoded in the same geometry. Your design here does enable easier backwards compatibility but also worries me about writing an efficient encoder/decoder and also would result in a slightly less compressed binary I would think.

Putting data into the 2D case could also be a problem for purely vertical lines in 3d for example. In this 2d case the X,Y would be a line-string that never changed and therefore would be violating the guarantees of linestrings as this would resolve to simply a point. Granted this is not a big concern but should be noted.

Reply:
### What unspecified elevation means

I agree that it's unreasonable to expect that encoders or decoders will know the ground elevation. But as far as I can tell, if no one knows what it is, the alternatives are:

* 0 and null both mean the ellipsoid, and 3-D features will tower above them
* 0 and null both mean the ground plane, and rendering 3-D geometries is impossible because the renderer also doesn't know the shape of the ground

Of these choices, I think the first is better, as unfortunate as it is, because I think the second is completely unworkable.

### One object or two

The main argument I can make in favor of a separate object for the third dimension is that we are contemplating at least one more pseudo-dimension (arbitrary attributes), and there has been a request for another (the [M-dimension](https://github.com/mapbox/vector-tile-spec/issues/76#issuecomment-405482889) from shapefile). If all of these dimensions live in the same object, then we either need 6 different versions of each operator or a bitfield that specifies all 6 ways to read each coordinate "pair".

* lineto (2-D)
* lineto (3-D)
* lineto (4-D)
* lineto (2-D with attributes)
* lineto (3-D with attributes)
* lineto (4-D with attributes)

My opinion is that giving each dimension its own object is both more compatible with the past (existing encoders and decoders do not need to change) and more future-proof because we can add whatever other parallel arrays may be necessary in the future as we think of other things that need to be encoded. Putting everything in the same object is neither compatible nor straightforwardly extensible.

You are right that we currently forbid 0-length movetos and linetos. Does anything actually enforce this or require it, or is it in the spec only as encouragement for people not to bloat their tiles with do-nothing geometry operations?